### PR TITLE
Scala 3: Macro migration

### PR DESCRIPTION
**Pull request by mkurz**

### CODE REVIEW

**Comment by jtjeferreira** [comment]
> I can't see the benefit of using val here

I think that was necessary before I moved the scala3 macros to `extension (inline sc: StringContext) {`. Nice that you spotted that is no longer necessary...

File: core/play/src/main/scala-3/play/api/routing/sird/package.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    /**
+     * String interpolator for extracting parameters out of URL paths.
+     *
+     * By default, any sub value extracted out by the interpolator will match a path segment, that is, any
+     * String not containing a /, and its value will be decoded.  If however the sub value is suffixed with *,
+     * then it will match any part of a path, and not be decoded.  Regular expressions are also supported, by
+     * suffixing the sub value with a regular expression in angled brackets, and these are not decoded.
+     */
+    val p: PathExtractor = PathExtractor.cached(sc.parts)
+  }
```
Comment (mkurz) [inline core/play/src/main/scala-3/play/api/routing/sird/package.scala:58]: I am not really happy that we now split the interpolators into the `UrlContext` implicit class and the extension methods... I would prefer to keep everything in `UrlContext` (which is not possible in Scala 3) or to also turn the `p` interpolator into an extension method... However, making `p` an extension method would mean it needs to be a method instead a `val` and I don't want to introduce any performance penalties (evaluating the cache each time the method gets called vs only doing that once when initializing the val), so it's ok for me to keep it that way.
Also given that in the docs we recommend to import `play.api.routing.sird._` it (hopefully) should not really matter.
We probably want to introduce a Scala 3 migration guide in the docs to mention things like that.

File: core/play/src/main/scala-2/play/api/mvc/PathBindableMacros.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+trait PathBindableMacros {
+  import scala.language.experimental.macros
+
+  /**
+   * Path binder for AnyVal
+   */
+  implicit def anyValPathBindable[T <: AnyVal]: PathBindable[T] =
+    macro BinderMacros.anyValPathBindable[T]
+}
```
Comment (mkurz) [inline core/play/src/main/scala-2/play/api/mvc/PathBindableMacros.scala:17]: Is tested here so we are good: https://github.com/playframework/playframework/blob/0cf20608958337481ff57c37c5d90ea130a799ef/core/play/src/test/scala/play/api/mvc/BindersSpec.scala#L246-L259

I also ran manual tests with the scripted tests modified in the PR: https://github.com/playframework/playframework/pull/10700 (also trying various things like removing anyval there, etc.)

File: core/play/src/main/scala-2/play/api/mvc/QueryStringBindableMacros.scala [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+package play.api.mvc
+
+import play.api.mvc.macros.BinderMacros
+
+trait QueryStringBindableMacros {
+  import scala.language.experimental.macros
+
+  implicit def anyValQueryStringBindable[T <: AnyVal]: QueryStringBindable[T] =
+    macro BinderMacros.anyValQueryStringBindable[T]
+}
```
Comment (mkurz) [inline core/play/src/main/scala-2/play/api/mvc/QueryStringBindableMacros.scala:14]: Is tested here, so we are good: https://github.com/playframework/playframework/blob/0cf20608958337481ff57c37c5d90ea130a799ef/core/play/src/test/scala/play/api/mvc/BindersSpec.scala#L261-L284

File: core/play/src/main/scala-2/play/api/routing/sird/macroimpl/QueryStringParameterMacros.scala [hunk 4]
```diff
@@ -38,6 +38,20 @@ private[sird] object QueryStringParameterMacros {
...
+          c.abort(
+            c.enclosingPosition,
+            "Invalid use of query string extractor with empty parts"
+          )
+        }
+
+        if (parts.sizeIs > 2) {
+          c.abort(
+            c.enclosingPosition,
+            "Query string extractor can only extract one parameter, extract multiple parameters using the & extractor, eg: " + name + "\"param1=$param1\" & " + name + "\"param2=$param2\""
```
Comment (mkurz) [inline core/play/src/main/scala-2/play/api/routing/sird/macroimpl/QueryStringParameterMacros.scala:51]: These compiler errors are not tested in the scripted test I added.

File: core/play/src/main/scala-3/play/api/mvc/macros/BinderMacros.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                    report.errorAndAbort(
+                      s"Instance not found: ${Type.show[T]}[${tpr.typeSymbol.fullName}]"
+                    )
+                }
+            }
+          }
+
+          case _ =>
+            report.errorAndAbort(
+              s"Constructor parameter expected, found: ${v}"
```
Comment (mkurz) [inline core/play/src/main/scala-3/play/api/mvc/macros/BinderMacros.scala:76]: IMHO (as far as I remember from last week) it's basically impossible to hit this and the below compiler error because of the nature of anyval (an AnyVal has a ValDef as constructor argument by defintion for example (if I got that right)

File: core/play/src/main/scala-3/play/api/routing/sird/package.scala [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  /**
+   * Same as &, but for convenience to make the dsl look nicer when extracting query strings
+   */
+  val ? = &
+
+  /**
+   * The query string type
+   */
+  type QueryString = Map[String, Seq[String]]
+}
```
Comment (mkurz) [inline core/play/src/main/scala-3/play/api/routing/sird/package.scala:121]: Alls interpolators here are tested in 
* https://github.com/playframework/playframework/blob/main/core/play/src/test/scala/play/api/routing/sird/DslSpec.scala and 
* https://github.com/playframework/playframework/blob/main/core/play-integration-test/src/it/scala/play/it/http/UriHandlingSpec.scala

**Comment by mkurz** [comment]
So, even if I only pushed some smaller commits here, I did more work and research (last week before I got sick) than it might look like: Macros have always been my blind spot in Scala and because I didn't want to hit the merge button without really understanding what's going on, it was time to dive in...

**TLDR: Let's merge the PR, it's ok, even not perfect, but no show stoppers.**

First, the scripted test I added makes sure that he compiler errors that get forced by the macros actually will occur...
I was adding that scripted test because when I was testing the PR I realized that when using things like `q"a=${a}b=${b}"` instead of the expected compiler error, I got:
```
Wrong number of argument patterns for _root_.scala.StringContext("a=", "b=", "").q; expected: (String)
```
That happend in Scala 3 only however.
It's known that in Scala 3 [things changed](https://docs.scala-lang.org/scala3/reference/changed-features/pattern-matching.html) in regards to `unapply` methods, so the easiest solution was to just switch to `unapplySeq` methods now, which I am not too happy about but it works for all cases. I don't like it too much because they don't really return a seq but always the first element... But anyway.
Actually, the reason `unapply` works in Scala 2 is because in Scala 2 the macro makes the whole compilation fail much earlier, so the compiler won't even try to "match" an `unapply[Seq]` method. That's because the Scala 2 macro work a bit different, by inspecting the actually code and tree and fails early if there are too many arguments passed to `StringContext.q`.
So at this point I got sucked in way too deep, because I wanted to achieve the same with Scala 3, to keep the `unapply` methods and not hit the above compiler error, but it seems that was not easily possible:
First I tried to keep the `QueryStringParameterExtractor`'s `unapply` methods and instead introduce `unapplySeq` methods directly on all the `StringContext.[q|q_o|...]` [extractors](https://docs.scala-lang.org/tour/extractor-objects.html) - just like Scala itself does with it's `StringContext.s` interpolator (see [here](https://github.com/scala/scala/blob/2.13.x/src/library/scala/StringContext.scala#L130)). Turns out the when using the [Scala 3 `extension`](https://docs.scala-lang.org/scala3/book/ca-extension-methods.html) approach that only starting with Scala 3.3.0-RC1 it was possible to define `unapply` methods on methods wrapped by the `extension`:
* https://github.com/lampepfl/dotty/issues/8577
* https://github.com/lampepfl/dotty/pull/16358

Thanks to that and [these tests](https://github.com/lampepfl/dotty/pull/16358/files#diff-7d4c46ab2cb40487d4e63816f21e5fc1edbc6fad5fa0945f93e5cd79fe8ead08) that were added to dotty finally I was able to add the `unapplySeq` methods to the extension methods... That was after I tried to still keep using the [`implicit class UrlContext`](https://docs.scala-lang.org/overviews/core/implicit-classes.html) instead of the `extension` approach, which does not work in Scala 3, but, like said I was a bit naive in regards to macros until now, I didn't know better yet...
Anyway, even though I was able to make that work, I couldn't really figure out how to come up with a general approach to "forward" the unapplySeq methods to the unapply methods of the `QueryStringParameterExtractor` without lots of duplicate code. So that's why I just stopped and kept this pull request like it is right now. I don't wanted to spend so much time with that... If someone wants to come up with a nicer solution, here is the "nice" version of my work (that was distributed in several branches) with some comments to make things a bit easier (there are other Scala 3 stuff like ["Opaque Type Aliases"](https://dotty.epfl.ch/docs/reference/other-new-features/opaques.html))

* https://github.com/mkurz/playframework/commits/macro_migration_avoid_unapplySeq ([diff](https://github.com/mkurz/playframework/commit/3d705df4cc9e0b2f1d64be93af2f1f4b88a4ceb8))

So anyway, there was other stuff coming up not mentioned here, besides reading lots of docs as well...
Like said I don't wanted to spend too much time with that macros stuff, but finally I can push the merge button with confidence.
